File: 在线仓库/modelzoo-master/samples/samples_GPL/built-in/yolov9s/script/yolov9s_preprocess.py
# Copyright (c) ModelZoo. 2025-2026. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import math
import cv2
import numpy as np
import argparse
import json
from tqdm import tqdm
from shutil import rmtree


def letterbox(img, target_size=(640, 640)):
    """保持比例缩放图片并填充至目标尺寸（与C++端LetterBox一致）"""
    shape = img.shape[:2]  # (h, w)

    # 计算缩放比例，不放大
    r = min(target_size[0] / shape[0], target_size[1] / shape[1])
    r = min(r, 1.0)

    # 计算未填充的尺寸
    new_unpad = (int(round(shape[1] * r)), int(round(shape[0] * r)))
    dw = (target_size[1] - new_unpad[0]) / 2.0
    dh = (target_size[0] - new_unpad[1]) / 2.0

    # 缩放
    if (shape[1], shape[0]) != new_unpad:
        img = cv2.resize(img, new_unpad, interpolation=cv2.INTER_LINEAR)

    # 填充
    top = int(round(dh - 0.1))
    bottom = int(round(dh + 0.1))
    left = int(round(dw - 0.1))
    right = int(round(dw + 0.1))
    img = cv2.copyMakeBorder(img, top, bottom, left, right,
                             cv2.BORDER_CONSTANT, value=[114, 114, 114])
    return img


def preprocess_image(image_path, target_size=(640, 640)):
    """预处理单张图片，与C++ Yolov9sPreprocess保持一致

    流程: 读取BGR -> 预缩放(长边=target) -> letterbox -> BGR2RGB -> /255.0 -> HWC->CHW
    """
    image = cv2.imread(image_path)
    if image is None:
        print(f"无法读取图片 {image_path}")
        return None

    h0, w0 = image.shape[:2]

    # 预缩放：等比缩放比取宽/高方向较小值，确保缩放后不超出目标框（与C++端一致）
    pre_r = min(target_size[0] / h0, target_size[1] / w0)
    if pre_r != 1.0:
        new_w = min(int(math.ceil(w0 * pre_r)), target_size[1])
        new_h = min(int(math.ceil(h0 * pre_r)), target_size[0])
        image = cv2.resize(image, (new_w, new_h), interpolation=cv2.INTER_LINEAR)

    # Letterbox填充
    image = letterbox(image, target_size)

    # BGR -> RGB
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)

    # 归一化到 [0, 1] float32（YOLOv9s不使用AIPP，需软件归一化）
    image = image.astype(np.float32) / 255.0

    # HWC -> CHW
    image = image.transpose(2, 0, 1)

    # 确保连续内存布局
    image = np.ascontiguousarray(image)

    # 添加batch维度 (1 x C x H x W)
    image = np.expand_dims(image, axis=0)

    return image


def image_to_bin(image_path, output_bin_path, target_size=(640, 640)):
    """将图片转换为bin文件

    Args:
        image_path: 输入图片路径
        output_bin_path: 输出bin文件路径
        target_size: 目标尺寸，默认为 (640, 640)

    Returns:
        bool: 处理是否成功
    """
    # 预处理图片
    processed_image = preprocess_image(image_path, target_size)
    if processed_image is None:
        return False

    # 保存为bin文件
    try:
        processed_image.tofile(output_bin_path)
        return True
    except Exception as e:
        print(f"保存bin文件失败 {output_bin_path}: {e}")
        return False


def process_images(file_list_path, output_dir, target_size=(640, 640), max_count=5000):
    """处理file_list.json中的图片，转换为bin文件并保存到输出文件夹

    Args:
        file_list_path: 包含图片路径的JSON文件
        output_dir: 输出目录
        target_size: 目标尺寸，默认为 (640, 640)
        max_count: 最大处理数量
    """
    # 如果目标目录已存在，先删除以清空旧数据
    if os.path.exists(output_dir):
        print(f"清空已存在的目录: {output_dir}")
        rmtree(output_dir)

    # 创建输出文件夹（如果不存在）
    os.makedirs(output_dir, exist_ok=True)

    # 读取file_list.json
    if not os.path.exists(file_list_path):
        print(f"错误: 文件列表 {file_list_path} 不存在")
        return

    with open(file_list_path, 'r') as f:
        try:
            data = json.load(f)
            image_files = [item[0] for item in data.get('fileList', [])]
        except json.JSONDecodeError as e:
            print(f"解析JSON文件失败: {e}")
            return

    # 限制处理数量
    process_files = image_files[:max_count]

    # 遍历并处理图片，带进度条
    success_count = 0
    for filename in tqdm(process_files, desc="YOLOv9s预处理进度"):
        # 检查文件是否存在
        if not os.path.exists(filename):
            print(f"文件不存在，跳过: {filename}")
            continue

        # 生成输出文件名（保持原名，替换扩展名为.bin）
        base_name = os.path.splitext(os.path.basename(filename))[0]
        output_path = os.path.join(output_dir, f"{base_name}.bin")

        # 转换图片为bin文件
        if image_to_bin(filename, output_path, target_size):
            success_count += 1

    print(f"处理完成，成功处理 {success_count}/{len(process_files)} 张图片")


def main():
    parser = argparse.ArgumentParser(description='YOLOv9s 数据预处理：将图片转换为float32 bin文件')

    parser.add_argument('--file_list', default='../data/file_list.json',
                        help='包含图片路径的JSON文件路径，默认../data/file_list.json')
    parser.add_argument('--output_dir', default='../out/preprocess/bin',
                        help='输出bin文件文件夹路径 (默认: ../out/preprocess/bin)')
    parser.add_argument('--target_size', type=int, nargs=2, default=[640, 640],
                        help='目标尺寸，格式为 高 宽，默认640 640')
    parser.add_argument('--max_count', type=int, default=5000,
                        help='最大处理图片数量，默认5000张')

    args = parser.parse_args()

    # 处理图片
    process_images(
        file_list_path=args.file_list,
        output_dir=args.output_dir,
        target_size=tuple(args.target_size),
        max_count=args.max_count
    )
    print(f"YOLOv9s预处理完成，bin文件已保存至 {args.output_dir}")


if __name__ == "__main__":
    main()

File: 在线仓库/modelzoo-master/samples/samples_GPL/common/infer/preprocess/yolov9s_preprocess.cpp
/*
 * Copyright (c) ModelZoo. 2025-2026. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "yolov9s_preprocess.h"
#include "log.h"
#include "utils.h"

#include <sys/stat.h>
#include <sys/types.h>
#include <algorithm>
#include <fstream>
#include <opencv2/opencv.hpp>
#include <string>
#include <vector>

static constexpr int BYTE_BIT_NUM = 8;
static constexpr int FP16_BIT_NUM = 16;
static constexpr int FP32_BIT_NUM = 32;
static constexpr float PIXEL_NORMALIZE_FACTOR = 255.0f;
static constexpr int CHANNEL_COUNT = 3;
static constexpr int PAD_COLOR_VAL = 114;
static constexpr mode_t FILE_PERMISSION_0755 = 0755;

namespace Infer {

// 预处理配置
struct PreprocessConfig {
    cv::Size imgSize;
    std::string saveBin;
    bool isDpico;
};

static void ConvertFp32ToFp16Inplace(const float* srcData, size_t dataCount, uint16_t* dstData)
{
    for (size_t i = 0; i < dataCount; ++i) {
        dstData[i] = FloatToHalf(srcData[i]);
    }
}

static std::vector<uint16_t> ConvertFp32ToFp16Vector(const float* srcData, size_t dataCount)
{
    std::vector<uint16_t> fp16Data(dataCount);
    ConvertFp32ToFp16Inplace(srcData, dataCount, fp16Data.data());
    return fp16Data;
}

static int64_t GetLoopTimes(const TensorDesc& desc)
{
    int64_t loopTimes = 1;
    for (size_t loop = 0; loop < desc.dimCount - 1; ++loop) {
        loopTimes *= desc.dims[loop];
    }
    return loopTimes;
}

static Result ReadImgToBufDpico(const cv::Mat& mat, const TensorDesc&,
    TensorBuf& inBuf)
{
    size_t matTotalBytes = mat.total() * mat.elemSize();
    memcpy(inBuf.GetRawPtr(), mat.data, matTotalBytes);
    return SUCCESS;
}

static Result ReadImgToBufDlite(const cv::Mat& mat, const TensorDesc& desc,
    TensorBuf& inBuf)
{
    const float* srcData = mat.ptr<float>();
    const bool isFp32 = (desc.typeSize == FP32_BIT_NUM);
    const bool isFp16 = (desc.typeSize == FP16_BIT_NUM);

    if (desc.defaultStride == 0) {
        if (isFp32) {
            memcpy(inBuf.GetRawPtr(), srcData, desc.defaultSize);
        } else {
            std::vector<uint16_t> fp16Data = ConvertFp32ToFp16Vector(srcData, mat.total());
            memcpy(inBuf.GetRawPtr(), fp16Data.data(), desc.defaultSize);
        }
        return SUCCESS;
    }

    int64_t loopTimes = GetLoopTimes(desc);
    int64_t width = desc.dims[desc.dimCount - 1];
    size_t lineSize = width * desc.typeSize / BYTE_BIT_NUM;
    size_t dstStride = inBuf.stride == 0 ? desc.defaultStride : inBuf.stride;
    char* dstBase = static_cast<char*>(inBuf.GetRawPtr());
    for (int64_t loop = 0; loop < loopTimes; loop++) {
        void* destPtr = dstBase + loop * dstStride;
        const float* srcPtr = srcData + loop * width;
        if (isFp32) {
            memcpy(destPtr, srcPtr, lineSize);
        } else {
            ConvertFp32ToFp16Inplace(srcPtr, width, static_cast<uint16_t*>(destPtr));
        }
    }
    return SUCCESS;
}

// Letterbox函数实现
static cv::Mat LetterBox(const cv::Mat& img, const cv::Size& target_size,
    bool scaleup)
{
    cv::Size shape = img.size();

    // 计算缩放比例
    double r = std::min(static_cast<double>(target_size.width) / shape.width,
        static_cast<double>(target_size.height) / shape.height);
    r = std::min(r, 1.0);

    // 计算未填充的尺寸
    cv::Size unpad(static_cast<int>(std::round(shape.width * r)),
        static_cast<int>(std::round(shape.height * r)));

    // 计算填充量
    double w = (target_size.width - unpad.width) / 2.0f;
    double h = (target_size.height - unpad.height) / 2.0f;
    cv::Mat resized;
    if (shape != unpad) {
        cv::resize(img, resized, unpad, 0, 0, cv::INTER_LINEAR);
    } else {
        resized = img.clone();
    }
    // 计算填充边界
    int top = static_cast<int>(std::round(h - 0.1));
    int bottom = static_cast<int>(std::round(h + 0.1));
    int left = static_cast<int>(std::round(w - 0.1));
    int right = static_cast<int>(std::round(w + 0.1));

    cv::Mat padded;
    cv::copyMakeBorder(resized, padded, top, bottom, left, right,
        cv::BORDER_CONSTANT, cv::Scalar(PAD_COLOR_VAL, PAD_COLOR_VAL, PAD_COLOR_VAL));

    return padded;
}

static cv::Mat HwcToChw(const cv::Mat& hwcImg)
{
    CV_Assert(!hwcImg.empty() && hwcImg.channels() == CHANNEL_COUNT);
    std::vector<cv::Mat> channels;
    cv::split(hwcImg, channels);

    cv::Mat chwMat(CHANNEL_COUNT, hwcImg.rows * hwcImg.cols, hwcImg.depth());
    chwMat = cv::Scalar(0);

    for (int channelIdx = 0; channelIdx < CHANNEL_COUNT; ++channelIdx) {
        cv::Mat flatChannel = channels[channelIdx].reshape(1, 1).clone();
        flatChannel.copyTo(chwMat.row(channelIdx));
    }

    return chwMat; // 返回 3 × (高度*宽度) 的CHW矩阵
}

// 提取文件名（不含路径和后缀）
static std::string GetFileName(const std::string& path) {
    size_t sep = path.find_last_of("/\\");
    size_t start = (sep == std::string::npos) ? 0 : sep + 1;
    size_t end = path.find_last_of(".");
    return (end == std::string::npos || end <= start) ? 
           path.substr(start) : path.substr(start, end - start);
}

// 预缩放：按比例缩放使长边等于目标尺寸
static void PreScaleImage(cv::Mat& img, const cv::Size& targetSize)
{
    int h0 = img.rows;
    int w0 = img.cols;
    // 等比缩放比：取宽/高方向的较小值，确保缩放后不超出目标框（preR == 1.0 时跳过）
    double preR = std::min(static_cast<double>(targetSize.width) / w0,
        static_cast<double>(targetSize.height) / h0);
    if (preR != 1.0) { // 整数相除，仅当维度完全相等时精确为 1.0，直接比较安全
        // ceil 后用 min 截断，防止浮点向上取整导致新尺寸超出目标框 1 个像素
        int newW = std::min(static_cast<int>(std::ceil(w0 * preR)), targetSize.width);
        int newH = std::min(static_cast<int>(std::ceil(h0 * preR)), targetSize.height);
        cv::resize(img, img, cv::Size(newW, newH), 0, 0, cv::INTER_LINEAR);
    }
}

// 图像变换流水线：预缩放 → letterbox → BGR2RGB → 归一化 → HWC转CHW
static cv::Mat TransformImage(cv::Mat& img, const cv::Size& targetSize)
{
    // 预缩放
    PreScaleImage(img, targetSize);

    // 应用letterbox
    cv::Mat processed = LetterBox(img, targetSize, false);

    // BGR到RGB：反转通道顺序
    cv::cvtColor(processed, processed, cv::COLOR_BGR2RGB);

    // 像素值归一化到 [0, 1] 区间
    cv::Mat floatImg;
    processed.convertTo(floatImg, CV_32FC3, 1.0 / PIXEL_NORMALIZE_FACTOR);

    // 将图像格式从 HWC 转换为 CHW
    cv::Mat chwImg = HwcToChw(floatImg);

    if (!chwImg.isContinuous()) {
        chwImg = chwImg.clone(); // 克隆矩阵以获得连续内存
    }
    return chwImg;
}

// 保存预处理结果为bin文件
static void SavePreprocessBin(const cv::Mat& chwImg, const std::string& binPath, const cv::Size& imgSize)
{
    std::ofstream binFile(binPath, std::ios::binary);
    if (binFile.is_open()) {
        size_t dataSize = imgSize.width * imgSize.height * CHANNEL_COUNT * sizeof(float);
        binFile.write((char*)chwImg.data, dataSize);
        binFile.close();
        chmod(binPath.c_str(), FILE_PERMISSION_0755);
        LOG(INFO) << "Saved preprocess bin: " << binPath;
    } else {
        LOG(WARNING) << "Failed to save preprocess bin: " << binPath;
    }
}

// 统一的数据写入派发
static void ReadImgToBuf(const cv::Mat& mat, const TensorDesc& desc, TensorBuf& buf, bool isDpico)
{
    if (isDpico) {
        ReadImgToBufDpico(mat, desc, buf);
    } else {
        ReadImgToBufDlite(mat, desc, buf);
    }
}

// 处理单张图像：读取 → 变换 → 写入buffer → 可选保存bin
static bool ProcessSingleImage(const std::string& imgPath, TensorDesc& desc,
    TensorBuf& buf, const PreprocessConfig& cfg)
{
    cv::Mat im0 = cv::imread(imgPath);
    if (im0.empty()) {
        LOG(ERROR) << "Read image failed: " << imgPath;
        return false;
    }

    cv::Mat chwImg = TransformImage(im0, cfg.imgSize);
    ReadImgToBuf(chwImg, desc, buf, cfg.isDpico);

    if (!cfg.saveBin.empty()) {
        std::string binPath = cfg.saveBin + "/" + GetFileName(imgPath) + ".bin";
        SavePreprocessBin(chwImg, binPath, cfg.imgSize);
    }
    return true;
}

bool Yolov9sPreprocess(std::vector<std::string>& fileList, std::vector<TensorBuf>& tensorBufs,
    std::vector<TensorDesc>& tensorDescs, bool isDpico)
{
    LOG(INFO) << "Detected " << (isDpico ? "dpico " : "dlite ");
    LOG(INFO) << "Processing file num: " << fileList.size();

    // 从cfg.txt读取配置
    auto cfgMap = ReadCfgFile("../data/cfg.txt");
    PreprocessConfig cfg;
    cfg.imgSize = cv::Size(std::stoi(cfgMap["img_width"]), std::stoi(cfgMap["img_height"]));
    cfg.saveBin = cfgMap["save_preprocess_bin"];
    cfg.isDpico = isDpico;

    // 按需创建目录
    if (!cfg.saveBin.empty() && !CreateDirectoryRecursive(cfg.saveBin)) {
        LOG(ERROR) << "Create preprocess bin dir failed";
        return false;
    }

    // 处理每个图像
    for (size_t i = 0; i < fileList.size(); ++i) {
        LOG(INFO) << "imgPath: " << fileList[i];
        if (!ProcessSingleImage(fileList[i], tensorDescs[i], tensorBufs[i], cfg)) {
            LOG(ERROR) << "PreProcessing failed!";
            return false;
        }
        LOG(INFO) << "read end: " << fileList[i];
    }

    LOG(INFO) << "PreProcessing completed successfully!";
    return true;
}

} // namespace Infer
